### PR TITLE
Fix/json parse

### LINKED ISSUE #66: Cannot parse variable with `=>`
We've encountered an issue when trying to send an email using a dynamic template, where one o the dynamic template variables contain a `=>`.

## How to reproduce

The following test case demonstrates the issue:

          it "should not mess up contents" do
            custom_args = {"text" => "line with a => in it"}
            mail["custom_args"] = custom_args
            mailer.deliver!(mail)
            expect(client.sent_mail['custom_args']).to eq({"text" => "line with a => in it"})
          end

The test case ends up failing with a 

    Failure/Error: JSON.parse(text, symbolize_names: symbolize)

     JSON::ParserError:
       767: unexpected token at '{"text":"line with "a":in it"}'
     # ./lib/sendgrid_actionmailer.rb:142:in `json_parse'
     # ./lib/sendgrid_actionmailer.rb:165:in `add_send_options'
     # ./lib/sendgrid_actionmailer.rb:33:in `deliver!'
     # ./spec/lib/sendgrid_actionmailer_spec.rb:323:in `block (5 levels) in <module:SendGridActionMailer>'

See https://github.com/eddiezane/sendgrid-actionmailer/pull/65/commits/2df06f5a5cfcc62ca9e9f4d355a5e43161dd017d to for detailed changes.

## Debugging

The immediate issue seems to be that `SendGridActionMailer::DeliveryMethod#json_parse` tries to convert the text representation of a Hash to JSON in order to parse it back to a Hash. In doing so, it changes all occurrences of `"=>"` to `":"`.

The issue seems to have been introduced in https://github.com/eddiezane/sendgrid-actionmailer/commit/413db5685f782547bf7db0f019d1749660d0fc10.

### PR DESCRIPTION
**Pull request by tyrauber**
Extends #67 to remove the json_parse method (and painful regex) in favor of using Mail 2.7 unparsed_value. Updates the transform_keys method with one based on Rails deep_transform_values which handles nested hashes. This version therefore is backwards compatible with previous rails versions.

Fixes #66. Closes #67.

### CODE REVIEW

File: lib/sendgrid_actionmailer.rb [hunk 1]
```diff
@@ -197,7 +185,7 @@ def add_personalizations(sendgrid_mail, mail)
       end
 
       if mail['personalizations']
-        personalizations = json_parse('[' + mail['personalizations'].value + ']', false)
+        personalizations = mail['personalizations'].respond_to?(:unparsed_value) ? mail['personalizations'].unparsed_value.to_a : mail['personalizations'].map(&:unparsed_value)
```
Comment (koppen) [inline lib/sendgrid_actionmailer.rb]: It's not entirely clear to me what's going on here; what is `mail["personalizations"]` expected to contain?

Would something like this do the same thing:

    personalizations = Array(mail["personalizations"]).map(&:unparsed_value)

?

File: lib/sendgrid_actionmailer.rb [hunk 2]
```diff
@@ -302,22 +291,17 @@ def perform_send_request(email)
...
-        block.call(key)
-      end if hash.respond_to?(:transform_keys)
-
-      hash.map do |key, value|
-        new_key = yield(key)
-        [new_key, value]
-      end.to_h
+    # Recursive key transformation based on Rails deep_transform_values
+    def self.transform_keys(object, &block)
+      #p object
```
Comment (koppen) [inline lib/sendgrid_actionmailer.rb]: Looks like a bit of debugging code snuck it's way in here :)

File: spec/lib/sendgrid_actionmailer_spec.rb [hunk 3]
```diff
@@ -341,100 +337,77 @@ def post(request_body:)
...
-            mailer.deliver!(mail)
-            expect(client.sent_mail['asm']).to eq({"group_id" => 99, "groups_to_display" => [4,5,6,7,8]})
-          end
-
-          it 'should not change values inside custom args' do
-            custom_args = { 'text' => 'line with a => in it' }
-            mail['custom_args'] = custom_args
-            mailer.deliver!(mail)
-            expect(client.sent_mail['custom_args']).to eq('text' => 'line with a => in it')
-          end
```
Comment (koppen) [inline spec/lib/sendgrid_actionmailer_spec.rb]: Sweet, red diff-lines :D

**Comment by koppen** [comment]
This is looking like a good step in the right direction, I'm liking it. Great seeing my changes being taken that step further :)

File: lib/sendgrid_actionmailer.rb (hunk 2 above)
Comment (tyrauber) [reply lib/sendgrid_actionmailer.rb]: Thanks, removed.

File: lib/sendgrid_actionmailer.rb (hunk 1 above)
Comment (tyrauber) [reply lib/sendgrid_actionmailer.rb]: This was due to the fact that mail['personalizations'] could have been an array of mail inputs, or a single mail input, but restructuring add_personalization to apply personalizations before mail attributes resolves this issue and cleans up the code further.

File: spec/lib/sendgrid_actionmailer_spec.rb (hunk 3 above)
Comment (tyrauber) [reply spec/lib/sendgrid_actionmailer_spec.rb]: Yeah, these use cases - 1.8 style hash and json hash - were not outlined in the Readme. Only supporting the 1.9 hash usage outlined in the Readme simplified the the code immensely. So axed.

**Comment by tyrauber** [comment]
Yeah, I really appreciate your contributions.  I am quite happy with the end result. I'll personally sleep better knowing that the json_parse and gsub regex are a thing of the past.  Let me know if you have any other questions or concerns. Otherwise I'll merge, cut 3.0 and update the Readme and changelog. Thanks again @koppen.

File: lib/sendgrid_actionmailer.rb [hunk 4]
```diff
@@ -75,33 +75,33 @@ def to_emails(input)
     end
 
     def setup_personalization(mail, personalization_hash)
-      p = Personalization.new
+      personalization = Personalization.new
```
Comment (koppen) [inline lib/sendgrid_actionmailer.rb]: Cheers for addressing a pet peeve of mine :D

File: lib/sendgrid_actionmailer.rb [hunk 5]
```diff
@@ -176,20 +176,18 @@ def add_content(sendgrid_mail, mail)
     end
 
     def add_personalizations(sendgrid_mail, mail)
+      if mail['personalizations']
+        mail['personalizations'].unparsed_value.each do |p|
+          sendgrid_mail.add_personalization(setup_personalization(mail, p))
+        end
+      end
```
Comment (koppen) [inline lib/sendgrid_actionmailer.rb]: This is way better 👍

**Pull request merged**